interp/declare.ml

```diff
@@ -453,28 +453,95 @@ let input_universe_context : universe_context_decl -> Libobject.obj =
+   bound at the *library* or *module* level. The polymorphic flag is
+   used to distinguish universes declared in polymorphic sections, which
+   are discharged and do not remain in scope. *)
+
+type universe_source =
+  | BoundUniv (* polymorphic universe, bound in a function (this will go away someday) *)
+  | QualifiedUniv of Id.t (* global universe introduced by some global value *)
+  | UnqualifiedUniv (* other global universe *)
+
+type universe_decl = universe_source * Nametab.universe_id
+
+let add_universe src (dp, i) =
+  let level = Univ.Level.make dp i in
+  let optpoly = match src with
+    | BoundUniv -> Some true
+    | UnqualifiedUniv -> Some false
+    | QualifiedUniv _ -> None
   in
-  cache_universe_context (p, ctx);
-  Global.set_global_universe_names glob'
+  Option.iter (fun poly ->
+      let ctx = Univ.ContextSet.add_universe level Univ.ContextSet.empty in
+      Global.push_context_set poly ctx;
+      Universes.add_global_universe level poly;
+      if poly then Lib.add_section_context ctx)
+    optpoly
 
-let input_universes : universe_decl -> Libobject.obj =
+let check_exists sp =
+  let depth = sections_depth () in
+  let sp = Libnames.make_path (pop_dirpath_n depth (dirpath sp)) (basename sp) in
+  if Nametab.exists_universe sp then
+    alreadydeclared (str "Universe " ++ Id.print (basename sp) ++ str " already exists")
+  else ()
+
+let qualify_univ src (sp,i as orig) =
+  match src with
+  | BoundUniv | UnqualifiedUniv -> orig
+  | QualifiedUniv l ->
+    let sp0, id = Libnames.repr_path sp in
+    let sp0 = DirPath.repr sp0 in
+    Libnames.make_path (DirPath.make (l::sp0)) id, i+1
+
+let cache_universe ((sp, _), (src, id)) =
+  let sp, i = qualify_univ src (sp,1) in
+  let () = check_exists sp in
+  let () = Nametab.push_universe (Nametab.Until i) sp id in
+    add_universe src id
+
+let load_universe i ((sp, _), (src, id)) =
+  let sp, i = qualify_univ src (sp,i) in
+  let () = Nametab.push_universe (Nametab.Until i) sp id in
+  add_universe src id
+
+let open_universe i ((sp, _), (src, id)) =
+  let sp, i = qualify_univ src (sp,i) in
+  let () = Nametab.push_universe (Nametab.Exactly i) sp id in
+  ()
+
+let discharge_universe = function
+  | _, (BoundUniv, _) -> None
+  | _, ((QualifiedUniv _ | UnqualifiedUniv), _ as x) -> Some x
+
+let input_universe : universe_decl -> Libobject.obj =
   declare_object
     { (default_object "Global universe name state") with
-      cache_function = (fun (na, pi) -> cache_universes pi);
-      load_function = (fun _ (_, pi) -> cache_universes pi);
-      discharge_function = (fun (_, (p, _ as x)) -> if p then None else Some x);
-      classify_function = (fun a -> Keep a) }
+      cache_function = cache_universe;
+      load_function = load_universe;
+      open_function = open_universe;
+      discharge_function = discharge_universe;
+      subst_function = (fun (subst, a) -> (** Actually the name is generated once and for all. *) a);
```

Comment:
@herbelin 
>By the way, how did you with functors, did you finally implement substitution, or is it that universes are regenerated at functor application?

I believe this line means neither.

Reply:
Indeed. Universes are not quantified at the functor level.

Reply:
If I do:
```coq
Module Type S. End S.
Module F(M:S). Definition u@{i} :=Type@{i}. End F.
Module M. End M.
Module N := F M.
Set Printing Universes.
Print N.u.
```
I see
```coq
N.u = Type@{N.u.i} : Type@{N.u.i+1}
(* {N.u.i} |=  *)
```
Who is giving the name `N.u.i`?

Reply:
Oh right the `a` is `Top.123`, the name is gotten from the libobject name thing (the id passed to `add_leaf`).

Reply:
OK, I think I got it (there is an object `universe` in the functor and the name comes from it).

Reply:
If you do:

Module N' := F M. (* N'.u.i is the same universes as N.u.i, nametab resolution now backwards resolves it to N'.u.i *)
Print N'.u.

Reply:
There seems to be a printing issue if you print the functor before constructing any instance. We could also refine this to not redeclare N.u.i and N'.u.i as aliases to the functors universe

Reply:
Generation of universes in functors is a bit strange for sure. If you like, the semantics of any (mono) universe declaration in a functor should be extruded to a toplevel declaration alongside the one of the functor that uses it.

Reply:
>Generation of universes in functors is a bit strange for sure. If you like, the semantics of any (mono) universe declaration in a functor should be extruded to a toplevel declaration alongside the one of the functor that uses it.

I'm not sure what that means exactly so maybe.

Reply:
If a universe in a functor is declared globally once for all for all instances of the functor, what about naming it using the name of the functor as prefix. Maybe it is not a good idea if we later consider functors which are generative on universes, but maybe that's acceptable in the current setting. Moreover, that would allow to use the name in this situation which currently does not use one:
```coq
Module Type S. End S.
Module F(M:S). Definition u@{i} :=Type@{i}. End F.
Set Printing Universes.
Print F. (* Module F := Functor (M:S) Struct Definition u : Type@{Top.1+1}. End *)
```
And similarly in this example where declaring a new module does a side-effect on the name of the universe declared in `F`:
```coq
Module Type S. End S.
Module F(M:S). Definition u@{i} :=Type@{i}. End F.
Module M. End M.
Set Printing Universes.
Module N := F M.
Print N.u. (* N.u = Type@{N.u.i} *)
Module N' := F M.
Print N.u. (* N.u = Type@{N'.u.i} *)
```
Or, maybe, we should simply refuse any declaration of universe name in functors (and hence in module types as well), so that there is no confusion on the status of these names, which morally are global. We would then force the users to explicitly declare these names at a global level.

Reply:
>If a universe in a functor is declared globally once for all for all instances of the functor, what about naming it using the name of the functor as prefix.

That's what I expected.

>Or, maybe, we should simply refuse any declaration of universe name in functors (and hence in module types as well), so that there is no confusion on the status of these names, which morally are global. We would then force the users to explicitly declare these names at a global level.

This seems like it could get irritating.

Reply:
> This seems like it could get irritating.

Then, we should probably do the same for module types.
```
Module Type S. Universe i. Definition A := Type@{i}. End S.
Set Printing Universes.
Print Module Type S.
```
should show
```coq
Module Type S = Sig Definition A : Type@{S.i+1}. End
```
and not
```coq
Module Type S = Sig Definition A : Type@{Top.1+1}. End
```
as the PR does.